File: src/pythonServer/starlette/starletteServer.ts
import { JupyterPackFramework } from '../../type';
import { BasePythonServer } from '../baseServer';

export class StarletteServer extends BasePythonServer {
  async init(options: {
    initCode?: string;
    instanceId: string;
    kernelClientId: string;
  }) {
    await super.init(options);
    const { initCode, instanceId, kernelClientId } = options;
    const baseURL = this.buildBaseURL({
      instanceId,
      kernelClientId,
      framework: JupyterPackFramework.STARLETTE
    });
    const bootstrapCode = `
    from jupyterpack.common import set_base_url_env
    set_base_url_env("${baseURL}")
    `;
    await this.kernelExecutor.executeCode({ code: bootstrapCode });
    if (initCode) {
      const initCodeWithUrl = initCode.replaceAll('{{base_url}}', baseURL);
      await this.kernelExecutor.executeCode({ code: initCodeWithUrl });
      const loaderCode = `
      from jupyterpack.asgi import AsgiServer
      ${this._server_var} = AsgiServer(app, "${baseURL}")
      `;

      await this.kernelExecutor.executeCode({ code: loaderCode });
    }
  }

  async disposePythonServer(): Promise<void> {
    await this.kernelExecutor.executeCode({
      code: `${this._server_var}.dispose()`
    });
    for (const element of this._openedWebsockets) {
      await this.closeWebsocket(element);
    }
  }

  async reloadPythonServer(options: {
    entryPath?: string;
    initCode?: string;
  }): Promise<void> {
    const { initCode } = options;
    if (initCode) {
      await this.kernelExecutor.executeCode({
        code: initCode.replaceAll('{{base_url}}', this._baseUrl ?? '')
      });
      const reloadCode = `
      await ${this._server_var}.dispose()
      ${this._server_var}.reload(app)
      `;
      await this.kernelExecutor.executeCode({ code: reloadCode }, true);
    }
  }
}

File: jupyterpack/starlette/starletteServer.py
from jupyterpack.asgi.asgiBridge import ASGIBridge
from ..asgi.asgiServer import AsgiServer
from starlette.applications import Starlette
from starlette.routing import Mount


class StarletteServer(AsgiServer):
    def __init__(self, starlette_app, base_url: str, origin: str):
        """
        Args:
            app : ASGI application instance
            base_url : base url for the application
        """

        main_app = Starlette(
            routes=[
                Mount(base_url, app=starlette_app),
            ]
        )
        super().__init__(main_app, base_url, origin)

    def reload(self, starlette_app):
        """
        Args:
            app : ASGIApplication instance
        """
        self._app = Starlette(
            routes=[
                Mount(self.base_url, app=starlette_app),
            ]
        )
        self._bridge = ASGIBridge(self._app, self.base_url, self._origin)
        return True
